FEATURE: Node copying Version 3 (as a Neos service)

mhsdesign opened the pull request:
Resolves: #5359 (because it will be obsolete)
Resolves: #5350
Resolves: #5351

**Upgrade instructions**

Instead of using `CopyNodesRecursively::createFromSubgraphAndStartNode` and handling that in the CR, please use:

```php
#[Flow\Inject()]
protected NodeDuplicationService $nodeDuplicationService;

// ...

$this->nodeDuplicationService->copyNodesRecursively(
    $contentRepositoryId,
    $workspaceName,
    $sourceDimensionSpacePoint,
    $sourceNodeAggregateId,
    $targetDimensionSpacePoint,
    $targetParentNodeAggregateId,
    $targetSucceedingSiblingNodeAggregateId
);
```



In case a leaf node that is tethered is copied, we un-tether it.
A migration `flow migrateevents:migratecopytetherednode` fixes this for previous cases.

Usages of the legacy copy node must be published before migrating to the final Neos version, to detect those see: `./flow migrateevents:copynodesstatus`

**Review instructions**

Neos UI part: https://github.com/neos/neos-ui/pull/3887

The with dd6408ecb8eab4778cdcff532f05b9b2e7122ba7 introduced `CopyNodesRecursively` content repository command is not stable, see its problems outline in #5351.

For one not all features were implemented (or only recently) like reference copying and subtree tag copying. Also there was little to no test coverage (2 tests)

Further by emitting events directly we would have to be super careful - which we were not. It was possible to copy tethered nodes and they kept their classification: #5350.

And lastly the event to command relation ship (needed for rebasing) is a problem. We cannot determine the copy command from the original event (which prevents us from removing the `commandClass` mapping).
And more importantly for rebasing the copy the WHOLE subtree snapshot is stored in the one events metadata field which can grow way to large (and is hard to maintain and write migrations for - everything is serialised in there). We argued which changes have to be done to keep copy nodes as a content repository command and found that there were almost no constraint checks for reapplying the copy (like the references of copied nodes could be deleted already), if we would check the constraints correctly, that would mean that the whole copy command could fail then which would not be desired. Instead, also to resolve the event to command relation ship we were thinking that a better low level copy implementation would generate the events with the command classes as if they were created by hand. Guessing the command payload (especially with the complexity of tethered nodes) would be utterly complex and errors would only be spottet during rebase. So we decided to actually issue REAL commands in a batch. As this is architecturally not possible from an actual cr command (would require full recursion) and that is okay, we discussed that we want a node copy service.
The disadvantages are that copying can fail now but thats also its positive side for rebasing. As a followup we need to find solutions to be graceful while copying and reapplying the copy. (Maybe by using `NodeAggregateCommandHandler::withoutAncestorNodeTypeConstraintChecks`)

And lastly, the copy command and its signature are not ideal as they dont allow multi dimension copying and also expose way to much internals: https://github.com/neos/neos-development-collection/issues/5359 This is no problem anymore by using this service.

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

Added Test-cases

- [x] test that the tethered state of the initial node is ignored
- [x] test that nested tethered nodes are copied WITH tethered state
  - [x] Even when its tethered directly in another thethered
- [x] test that tethered state of child nodes is redetermined as the parent node is now detached so a second level tethered node might not be tethered anymore, actually forbid that case
- [x] copying disabled nodes
- [x] root nodes cannot be copied
- [x] recursive copying, with properties, with references

mhsdesign committed:
WIP: Node Copy Version 3, first draft of NodeDuplicationService

mhsdesign committed:
WIP: Fix copy tethered node if leave node or nested

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/copy-nodes-as-a-service

mhsdesign committed:
BUGFIX: Only leaf tethered nodes can be copied

mhsdesign committed:
TASK: Reintroduce reference copying for first node

mhsdesign committed:
TASK: Add further testcases for node copy

mhsdesign committed:
TASK: Test and fix recursive node copying

mhsdesign committed:
TASK: Add test for copying tethered nodes deep in the tree

mhsdesign committed:
TASK: Add failing test for copying nested references

mhsdesign committed:
TASK: Add failing test for copying nested tethered nodes

mhsdesign committed:
BUGFIX: Fix copying nested tethered nodes

mhsdesign committed:
BUGFIX: Copy references for child nodes on copy

mhsdesign committed:
TASK: Remove unused constraint check helpers from `TransientNodeCopy`

mhsdesign committed:
TASK: Move `NodeAggregateIdMapping` to Neos as it will be removed from the cr core soon

mhsdesign committed:
TASK: Make phpstan happy

mhsdesign committed:
TASK: Fix specifying `references` in a row

mhsdesign committed:
TASK: Add proper constraint-check exceptions

mhsdesign committed:
TASK: Remove todo regarding additional property values for first node

the idea was that command hooks must be able to set a unique uri path segment, but as were using now a high level service we dont need to intercept that.

kitsunet commented on Neos.Neos/Classes/Domain/Service/NodeDuplication/Commands.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * @implements \IteratorAggregate<CommandInterface>
+ */
+final readonly class Commands implements \IteratorAggregate, \Countable
```
I don't really like this here, it seems confusing that there is no generic Commands collection class in the CR but then there is one here in Neos NodeDuplication. Anyone searching for somehting like this will trip over it...

mhsdesign replied:
agreed, now its in the core :)

mhsdesign committed:
TASK: (re)Introduce `Subtrees` as typed array for `$children`

Related https://github.com/neos/neos-development-collection/issues/4535

mhsdesign committed:
!!! TASK: Soft removal of `CopyNodesRecursively`

The command `CopyNodesRecursively` still needs to preserved as `RebasableToOtherWorkspaceInterface` "internal" command so publishing from user workspaces continues to work.

mhsdesign committed:
TASK: Make `NodeAggregateIdMapping` usable and API, as its used in the Neos ui to fetch the newly create node

mhsdesign committed:
TASK: Remove `$targetNodeName` from first level api option in `copyNodesRecursively`

as node names are deprecated.

Instead we should rather expose the to be run `Commands` and then one can change the node name

mhsdesign committed:
TASK: Move generic `Commands` into cr core

mhsdesign committed:
TASK: Document `NodeDuplicationService`

mhsdesign committed:
FEATURE: Copy subtree tags during copy

mhsdesign commented on Neos.Neos/Classes/Domain/Service/NodeDuplicationService.php:
```diff
@@ -0,0 +1,265 @@
+ * @Flow\Scope("singleton")
+ * @api
+ */
+final class NodeDuplicationService
```
to support usecases like https://github.com/Flowpack/Flowpack.NodeTemplates/issues/60 were we need to return a list of commands and not handle them ourselves, it would be practical to expose a way to get the commands we would be handle but dont handle them.

Like: 

```php
public function calculateCopyNodesRecursively(
    ContentRepositoryId $contentRepositoryId,
    WorkspaceName $workspaceName,
    DimensionSpacePoint $sourceDimensionSpacePoint,
    NodeAggregateId $sourceNodeAggregateId,
    OriginDimensionSpacePoint $targetDimensionSpacePoint,
    NodeAggregateId $targetParentNodeAggregateId,
    ?NodeAggregateId $targetSucceedingSiblingNodeAggregateId,
    ?NodeAggregateIdMapping $nodeAggregateIdMapping = null
): Commands;
```

i would declare this as experimental api but it would be helpful to have that. As otherwise node copying would not be possible there without duplicating this.

kitsunet replied:
Sounds reasonable to me

mhsdesign committed:
BUGFIX: Migration to fix previously copied tethered nodes #5350

In case a leaf node that is tethered is copied, we un-tether it.
A migration `flow migrateevents:migratecopytetherednode` fixes this for previous cases.

mhsdesign commented:
The `CopyNodesRecursively` command will be removed with the release of neos 9, but is still present to allow the publication of workspaces. A warning is emitted in case workspaces contain changes.

```
migrateevents:copyNodesStatus

WARNING: 1 content streams contain unpublished legacy copy node events. They MUST be published before migrating to Neos 9 and will not be publishable afterwards.
Content stream e76a0472-0c6f-4b98-bfac-12c7a2515dfb (admin-admington-1)
  - @1093 copy node 55861d58-c4e5-6121-b7cf-ef9320ebfbb0 to 55861d58-c4e5-6121-b7cf-ef9320ebfbb0
NOTE: To reduce the number of "false positives" above and to cleanup the event store run `./flow contentStream:removeDangling` and `./flow contentStream:pruneRemovedFromEventStream`
```

mhsdesign committed:
TASK: Introduce `migrateevents:copyNodesStatus`

kitsunet commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -304,7 +305,7 @@ public function findSubtree(NodeAggregateId $entryNodeAggregateId, FindSubtreeFi
                 $this->visibilityConstraints
             );
-            $subtree = new Subtree((int)$nodeData['level'], $node, array_key_exists($nodeAggregateId, $subtreesByParentNodeId) ? array_reverse($subtreesByParentNodeId[$nodeAggregateId]) : []);
+            $subtree = Subtree::create((int)$nodeData['level'], $node, Subtrees::fromArray(array_key_exists($nodeAggregateId, $subtreesByParentNodeId) ? array_reverse($subtreesByParentNodeId[$nodeAggregateId]) : []));
```
took me a moment here, I think I might prefer a Subtrees::empty() and then `array_key_exists($nodeAggregateId, $subtreesByParentNodeId) ? Subtrees::fromArray(array_reverse($subtreesByParentNodeId[$nodeAggregateId])) : Subtrees:empty()`

kitsunet reviewed (approved):
Provisional +1, looks already pretty good, no major complaints atm.

mhsdesign committed:
TASK: Introduce `Subtrees::createEmpty`

mhsdesign committed:
TASK: Adjust to new behat gherkin version to use quadruple back-slash

mhsdesign committed:
TASK: Add test for simple constraint checks

mhsdesign committed:
TASK: Introduce internal `calculateCopyNodesRecursively`

to support usecases like https://github.com/Flowpack/Flowpack.NodeTemplates/issues/60 were we need to return a list of commands and cant handle them ourselves, it would be practical to expose a way to get the commands that would be handled

mhsdesign commented:
In a followup we could improve things and address the optional todos

> For example copying a node with tethered children that is not tethered according to the current node type schema, or copying properties that are not defined
in the current node type schema anymore. In those cases the structure adjustments have to be executed. (todo only copy what is applicable and be graceful)

In Flowpack.NodeTemplates i already build a graceful mechanism by carefully checking all constraints beforehand. This is of-course tedious but might be worth in in a later point:

https://github.com/Flowpack/Flowpack.NodeTemplates/releases/tag/2.0.0

And also add tests like this:

```
Scenario: Coping fails if a node with tethered children is copied that are not tethered according to the current node type schema
  When I change the node types in content repository "default" to:
  """yaml
  'Neos.ContentRepository.Testing:TetheredDocument':
    childNodes: {} # no tethered node anymore!
  'Neos.ContentRepository.Testing:DocumentWithoutTetheredChildren': []
  """

  And I expect the node aggregate "davids-tether" to exist
  And I expect this node aggregate to be classified as "tethered"

  When copy nodes recursively is executed with payload:
    | Key                         | Value                 |
    | sourceDimensionSpacePoint   | {}                    |
    | sourceNodeAggregateId       | "sir-david-nodenburg" |
    | targetDimensionSpacePoint   | {}                    |
    | targetParentNodeAggregateId | "node-mc-nodeface"    |
```

mhsdesign commented:
Hmm i just noticed that copying behaved differently in Neos 8.3 :O

As per https://github.com/neos/neos-development-collection/blob/b03cca1eafc3247e17b7f2a6a3b3deaa5a4eba9e/Neos.ContentRepository/Classes/Domain/Model/NodeType.php#L283

> Recursive copying only happens *inside* this aggregate, and stops at nested aggregates.

Meaning that when copying a document only its content children are copied but not nested documents? That could be implemented now with a `Neos.Neos:Content` filter on the subtree... And this would be even fully okay as this is a Neos Service now.

But it makes me wonder if we really should remove the `isAggregate` as that would mean we would never be able to have this 1-to-1 copy behaviour in the core as the core doesnt know anything about content vs document.

mhsdesign committed:
TASK: Rename `NodePath::forRoot` to `NodePath::createEmpty()`

Because ROOT paths are only a concept of `AbsoluteNodePath` and not by a singular `NodePath` itself.

mhsdesign committed:
TASK: Remove obsolete logic to keep tethered node ids

kitsunet commented:
> Recursive copying only happens inside this aggregate, and stops at nested aggregates.

This is a lie though, I see no such check in Node::createRecursiveCopy and am not aware of one... 😜

mhsdesign commented:
Jup tested it in 8.3 it doesnt do that :D

But we actually used it as `$detachedCopy` flag:

https://github.com/neos/neos-development-collection/blob/6936b7bfab084027df2e6c12d01944a1ef7c5ba8/Neos.ContentRepository/Classes/Domain/Model/Node.php#L812

> $detachedCopy only has an influence if we are copying from one dimension to the other, possibly creating a new node variant

see https://github.com/neos/neos-development-collection/blob/6936b7bfab084027df2e6c12d01944a1ef7c5ba8/Neos.ContentRepository/Classes/Domain/Model/Node.php#L1670-L1681

so not interestring for our case as copying for dimensions is kinda broken still :D -> https://github.com/neos/neos-development-collection/issues/5054
